bridges/snowbridge/parachain/pallets/inbound-queue/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+
+			// Retrieve the registered channel for this message
+			let channel =
+				T::ChannelLookup::lookup(envelope.channel_id).ok_or(Error::<T>::InvalidChannel)?;
+
+			// Verify message nonce
+			<Nonce<T>>::try_mutate(envelope.channel_id, |nonce| -> DispatchResult {
+				if *nonce == u64::MAX {
+					return Err(Error::<T>::MaxNonceReached.into())
+				}
+				if envelope.nonce != nonce.saturating_add(1) {
+					Err(Error::<T>::InvalidNonce.into())
+				} else {
+					*nonce = nonce.saturating_add(1);
+					Ok(())
+				}
+			})?;
+
+			// Reward relayer from the sovereign account of the destination parachain
+			// Expected to fail if sovereign account has no funds
+			let sovereign_account = sibling_sovereign_account::<T>(channel.para_id);
+			let delivery_cost = Self::get();
+			T::Token::transfer(&sovereign_account, &who, delivery_cost, Preservation::Preserve)?;
+
+			// Decode message into XCM
+			let (xcm, fee) =
+				match inbound::VersionedMessage::decode_all(&mut envelope.payload.as_ref()) {
+					Ok(message) => Self::do_convert(envelope.message_id, message)?,
+					Err(_) => return Err(Error::<T>::InvalidPayload.into()),
+				};
+
+			// We embed fees for xcm execution inside the xcm program using teleports
+			// so we must burn the amount of the fee embedded into the XCM script.
+			T::Token::burn_from(&sovereign_account, fee, Precision::Exact, Fortitude::Polite)?;
+
+			log::info!(
+				target: LOG_TARGET,
+				"💫 xcm {:?} sent with fee {:?}",
+				xcm,
+				fee
+			);
+
+			// Attempt to send XCM to a dest parachain
+			let message_id = Self::send_xcm(xcm, channel.para_id)?;
+
+			Self::deposit_event(Event::MessageReceived {
+				channel_id: envelope.channel_id,
+				nonce: envelope.nonce,
+				message_id,
+			});
+
+			Ok(())
+		}
+
+		/// Halt or resume all pallet operations. May only be called by root.
+		#[pallet::call_index(1)]
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operating_mode(
+			origin: OriginFor<T>,
+			mode: BasicOperatingMode,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			OperatingMode::<T>::set(mode);
+			Self::deposit_event(Event::OperatingModeChanged { mode });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_convert(
+			message_id: H256,
+			message: inbound::VersionedMessage,
+		) -> Result<(Xcm<()>, BalanceOf<T>), Error<T>> {
+			let (mut xcm, fee) =
+				T::MessageConverter::convert(message).map_err(|e| Error::<T>::ConvertMessage(e))?;
+			// Append the message id as an XCM topic
+			xcm.inner_mut().extend(vec![SetTopic(message_id.into())]);
+			Ok((xcm, fee))
+		}
+
+		pub fn send_xcm(xcm: Xcm<()>, dest: ParaId) -> Result<XcmHash, Error<T>> {
+			let dest = MultiLocation { parents: 1, interior: X1(Parachain(dest.into())) };
+			let (xcm_hash, _) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
+			Ok(xcm_hash)
+		}
+	}
+
+	/// API for accessing the delivery cost of a message
```

Comment:
> delivering message also requires delivery of some header/header proof (i.e. some call of the snowbridge-ethereum-beacon-client pallet), right? In your scheme, you are not rewarding relayers for submitting those headers, just for delivering messages

Correct, delivery of consensus messages (headers) is not incentivized, the plan is for consensus relayers to be funded via the treasury.

> we've had a following problem with separate submit-header and submit-message transactions

Yeah, that's one of the reasons why we choose to have totally different funding mechanisms for consensus relayers and message relayers, treasury funding and per-message rewards respectively.

> this amount, returned by get() is a cost of execution, but it doesn't include the cost of transaction bytes .

Good catch! We overlooked this, and I think we've already pushed a fix into this PR.

The `Message` object makes up the bulk of the transaction data, and according to our analysis, will at most be 2KiB, which unfortunately will mean an extra $14 length fee for users at current DOT prices.

Reply:
> Correct, delivery of consensus messages (headers) is not incentivized, the plan is for consensus relayers to be funded via the treasury.

Ok, thank you! I have a few more questions here:
- do we expect every ethereum header to be relayed to the Polkadot BH (or here: Rococo BH)?
- if no, is there some limits on that?

I'm mostly worrying about the case when dumb/malicious consensus relayer starts submitting every eth header, occupying the whole bridge hub space (`ref_time` and `proof_size`), thus pushing e.g. our bridge (Polkadot <> Kusama) transactions out. And if he gets refunded for that, then why not?

In other words - if consensus relayers have been offline for some time - let's say a day, which is roughly `24 * 60 * 60 / 15 = 5_760` eth headers - can someone just start submitting those `5_760` headers? Can the "honest" relayer just submit some last header to stop this flood?

If the answer is yes, then I'm just fine with that :)

> The Message object makes up the bulk of the transaction data, and according to our analysis, will at most be 2KiB, which unfortunately will mean an extra $14 length fee for users at current DOT prices.

Yeah, size component affect cost a lot. Thanks for fixing - will check it out!

Reply:
> do we expect every ethereum header to be relayed to the Polkadot BH (or here: Rococo BH)?

@svyatonik yes, we expect every ethereum header to be relayed to BH. This is by design. We could in the future change the light client to only sync every few ethereum headers (and those containing messages), but at the moment we sync all ethereum headers in order. We do not store the full ethereum header, [only a few fields](https://github.com/paritytech/polkadot-sdk/blob/d966d4628d12511d4d5fcd326ffcd9f99f63db89/bridges/snowbridge/parachain/primitives/beacon/src/types.rs#L353) that we are interested in.

> In other words - if consensus relayers have been offline for some time - let's say a day, which is roughly 24 * 60 * 60 / 15 = 5_760 eth headers - can someone just start submitting those 5_760 headers? 

This is the expected behaviour if the consensus relayer goes offline. The headers should sync fairly quickly, roughly speaking we've seen a week's (which is much longer than we expect to see in prod) backlog of headers synced in a few hours. @alistair-singh correct me if I'm wrong here. 😄

Reply:
Ok, thank you for the info!

The fact that the bridge hub may stay congested for hours (or a few hours)  is a worrisome, though. The `pallet-transaction-payment` [fee factor](https://github.com/paritytech/polkadot-sdk/blob/c9f9b4b2a84ba2578f64e39bbdb4c9499792bc92/substrate/frame/transaction-payment/src/lib.rs#L397C10-L400) will go "to the moon" in this case, which would make BridgeHub transactions much more expensive. All our fees won't be able to cover this extra congestion cost (because they're paid at other chains, which do not know current congestion factor).

Reply:
Ok, maybe "to the moon" isn't the right word here :) But fees will definitely increase - we (I?) need to check later how much your `submit` transactions fits the block and how it'll change fees at Polkadot BH.

Also - you've said that the "consensus relayers to be funded via the treasury". Is that correct, that you're going to run single instance of the relayer, right? And those funds will be asked in advance - e.g. for next year? Asking because if e.g. relayers will be refunded post-factum and/or if there'll be two concurrent relayers, then it may submit the same headers, which would cause x2 block occupation. And if they'll be refunded for that, it'd be bad.

I'm paying attention to all this stuff ^^^, because it also affects Polkadot <> Kusama bridge as well. That's where we collide :)

Reply:
@svyatonik these are valid points, thank you for raising them. We'll do 2 things here:

1. Throttle execution header updates based on fee factor: As an off-chain safeguard, we will throttle syncing ethereum headers when the fee factor on BH reaches a certain threshold. I have added an internal team ticket for it here: [Throttle execution header updates based on fee factor](https://linear.app/snowfork/issue/SNO-808) The current setup is that the relayer syncs 8 execution headers at a time (send 8 separate extrinsic calls, and waits for all of them to complete before sending more). We will throttle that down even more when the fee factor increases.
2. Only sync relevant execution header: Currently, we sync execution headers in order. However, since the execution headers are just used to verify messages and do not form part of the ethereum consensus, we could only import execution headers that actually contain messages. This is a larger on-chain change that we'll do later. I've created internal issue [Only import execution headers that contain messages](https://linear.app/snowfork/issue/SNO-809)